File: app/ios/Runner/OpenWithChannel.swift
import Flutter
import UIKit
import UniformTypeIdentifiers

/// The iOS half of the `mdviewer/openwith` platform channel — delivers a
/// single "Open with MDViewer" file (Info.plist's `CFBundleDocumentTypes` +
/// `LSSupportsOpeningDocumentsInPlace`) from either a cold launch
/// (`SceneDelegate.scene(_:willConnectTo:options:)`'s `connectionOptions
/// .urlContexts`) or a warm delivery while the app is already running
/// (`SceneDelegate.scene(_:openURLContexts:)`). See
/// `app/lib/src/vault/open_with_channel.dart` for the Dart-side contract:
/// `takeInitialFile` (cold, pulled once at startup) / `onOpenFile` (warm,
/// pushed).
///
/// Cold vs. warm is arbitrated with a tiny explicit handshake rather than
/// relying on undocumented platform-channel message buffering: the very
/// first Dart-side call is always `takeInitialFile`, which both consumes
/// whatever arrived before Dart was ready (`pendingFile`) and flips
/// `dartReady`. Anything delivered after that flip is pushed straight to
/// Dart via `onOpenFile` instead of being buffered for a
/// `takeInitialFile` call that will never come again.
///
/// This channel's actual open-URL delivery only runs on a real device or
/// simulator receiving a genuine `UIOpenURLContext` — it is device-E2E
/// verified (Task 8), not covered by `flutter test` (see `VaultChannel`'s
/// doc comment for the same caveat on the sibling `mdviewer/vault`
/// channel).
final class OpenWithChannel: NSObject {
  static let channelName = "mdviewer/openwith"

  private let channel: FlutterMethodChannel
  private var pendingFile: [String: Any]?
  private var dartReady = false

  init(messenger: FlutterBinaryMessenger) {
    channel = FlutterMethodChannel(name: Self.channelName, binaryMessenger: messenger)
    super.init()
    channel.setMethodCallHandler { [weak self] call, result in
      self?.handle(call, result: result)
    }
  }

  private func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    switch call.method {
    case "takeInitialFile":
      dartReady = true
      let file = pendingFile
      pendingFile = nil
      result(file)
    default:
      result(FlutterMethodNotImplemented)
    }
  }

  /// Called by `SceneDelegate` for both cold-start and warm delivery — a
  /// security-scoped, "open in place" URL per
  /// `LSSupportsOpeningDocumentsInPlace`. Reads the file's bytes while the
  /// security scope is held, then releases it immediately: v1's open-with
  /// flow is a one-shot read, not a held grant (unlike the folder-vault
  /// bookmarks `VaultChannel` persists) — see task-7-report.md's "relative
  /// images decline for open-with docs" limitation, which follows directly
  /// from not keeping this scope (or any folder context) around.
  func handleOpen(url: URL) {
    let scoped = url.startAccessingSecurityScopedResource()
    defer { if scoped { url.stopAccessingSecurityScopedResource() } }

    guard let data = try? Data(contentsOf: url) else { return }
    let payload: [String: Any] = [
      "name": url.lastPathComponent,
      "mimeType": mimeType(for: url),
      "bytes": FlutterStandardTypedData(bytes: data),
    ]

    if dartReady {
      channel.invokeMethod("onOpenFile", arguments: payload)
    } else {
      pendingFile = payload
    }
  }

  private func mimeType(for url: URL) -> String {
    guard let type = UTType(filenameExtension: url.pathExtension) else { return "text/markdown" }
    return type.preferredMIMEType ?? "text/markdown"
  }
}

File: app/android/app/src/main/kotlin/com/sriannamalai/app/OpenWithChannel.kt
package com.sriannamalai.app

import android.app.Activity
import android.content.Intent
import android.net.Uri
import android.provider.OpenableColumns
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel

/**
 * The Android half of the `mdviewer/openwith` platform channel — delivers a
 * single "Open with MDViewer" file from the manifest's `VIEW` intent-filters
 * (`AndroidManifest.xml`), read via [android.content.ContentResolver]. See
 * `app/lib/src/vault/open_with_channel.dart` for the Dart-side contract:
 * `takeInitialFile` (cold, pulled once at startup) / `onOpenFile` (warm,
 * pushed).
 *
 * Cold vs. warm is arbitrated the same explicit handshake the iOS side
 * (`ios/Runner/OpenWithChannel.swift`) uses rather than relying on
 * undocumented platform-channel message buffering: the first Dart-side call
 * is always `takeInitialFile`, which both consumes whatever arrived before
 * Dart was ready ([pendingFile]) and flips [dartReady]. Anything delivered
 * after that flip is pushed straight to Dart via `onOpenFile`.
 *
 * [MainActivity] calls [handleIntent] twice: once in `configureFlutterEngine`
 * with its own launch `intent` (cold start — a `VIEW` intent *is* how the
 * activity got created), and again from `onNewIntent` (warm — `singleTop`
 * launch mode means the same activity instance receives a second `VIEW`
 * intent instead of a new one being created, which is exactly what lets a
 * long-lived engine/channel handle it without a relaunch).
 *
 * This channel's actual intent-delivery behavior only runs on a real device
 * or emulator receiving a genuine `VIEW` intent — it is device-E2E verified
 * (Task 8), not covered by `flutter test` (same caveat `VaultChannel`'s doc
 * comment notes for the sibling `mdviewer/vault` channel).
 */
class OpenWithChannel(private val activity: Activity, messenger: BinaryMessenger) : MethodChannel.MethodCallHandler {
  companion object {
    private const val CHANNEL_NAME = "mdviewer/openwith"
  }

  private val channel = MethodChannel(messenger, CHANNEL_NAME)
  private var pendingFile: Map<String, Any?>? = null
  private var dartReady = false

  init {
    channel.setMethodCallHandler(this)
  }

  override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
    when (call.method) {
      "takeInitialFile" -> {
        dartReady = true
        val file = pendingFile
        pendingFile = null
        result.success(file)
      }
      else -> result.notImplemented()
    }
  }

  /** No-op for anything that isn't a `VIEW` intent carrying a data [Uri] — safe to call unconditionally from both call sites. */
  fun handleIntent(intent: Intent?) {
    if (intent == null || intent.action != Intent.ACTION_VIEW) return
    val uri = intent.data ?: return
    val payload = readFile(uri) ?: return

    if (dartReady) {
      channel.invokeMethod("onOpenFile", payload)
    } else {
      pendingFile = payload
    }
  }

  private fun readFile(uri: Uri): Map<String, Any?>? {
    val resolver = activity.contentResolver
    val bytes = runCatching { resolver.openInputStream(uri)?.use { it.readBytes() } }.getOrNull() ?: return null
    val name = queryDisplayName(uri) ?: uri.lastPathSegment ?: "Untitled.md"
    val mimeType = resolver.getType(uri) ?: "text/markdown"
    return mapOf("name" to name, "mimeType" to mimeType, "bytes" to bytes)
  }

  /** `content://` URIs commonly expose their human-readable filename only via [OpenableColumns.DISPLAY_NAME], not the URI's own last path segment (often an opaque document ID). */
  private fun queryDisplayName(uri: Uri): String? =
    runCatching {
      activity.contentResolver.query(uri, arrayOf(OpenableColumns.DISPLAY_NAME), null, null, null)?.use { cursor ->
        if (!cursor.moveToFirst()) return@use null
        val index = cursor.getColumnIndex(OpenableColumns.DISPLAY_NAME)
        if (index < 0) null else cursor.getString(index)
      }
    }.getOrNull()
}
